src/core/storage/fileio/s3_filesys.hpp

```diff
@@ -0,0 +1,384 @@
+class SeekStream : public Stream {
+ public:
+  // virtual destructor
+  virtual ~SeekStream(void) {}
+  /*! \brief seek to certain position of the file */
+  virtual void Seek(size_t pos) = 0;
+
+  /*! \brief tell the position of the stream */
+  virtual size_t Tell(void) = 0;
+
+  /*! tell the physical size of the stream */
+  virtual size_t FileSize(void) const = 0;
+
+  /*! \brief Returns true if at end of stream */
+  virtual bool AtEnd(void) const = 0;
+  /*!
+   * \brief generic factory function
+   *  create an SeekStream for read only,
+   *  the stream will close the underlying files upon deletion
+   *  error will be reported and the system will exit when create failed
+   * \param uri the uri of the input currently we support
+   *            hdfs://, s3://, and file:// by default file:// will be used
+   * \param allow_null whether NULL can be returned, or directly report error
+   * \return the created stream, can be NULL when allow_null == true and file do
+   * not exist
+   */
+  static SeekStream *CreateForRead(const char *uri, bool allow_null = false);
+};
+
+/*!
+ * \brief reader stream that can be used to read from AWS SDK
+ */
+class AWSReadStreamBase : public SeekStream {
+ public:
+  virtual ~AWSReadStreamBase() {
+    logstream(LOG_DEBUG) << "~AWSReadStream" << std::endl;
+    Close();
+  }
+
+  virtual void Close() {
+    logstream(LOG_DEBUG) << "AWSReadStream::Close()" << std::endl;
+    Reset(file_size_);
+  }
+
+  virtual size_t Tell(void) { return curr_bytes_; }
+
+  virtual size_t FileSize() const { return file_size_; }
+
+  virtual bool AtEnd(void) const { return curr_bytes_ == file_size_; }
+
+  virtual void Write(const void *ptr, size_t size) {
+    std_log_and_throw(std::runtime_error,
+                      "AWSReadStreamBase is not supposed to write");
+  }
+  // lazy seek function
+  virtual void Seek(size_t pos) {
+    ASSERT_TRUE(pos < file_size_);
+    if (curr_bytes_ != pos) {
+      this->Reset(pos);
+    }
+  }
+  virtual size_t Read(void *ptr, size_t size);
+
+ protected:
+  AWSReadStreamBase() : file_size_(0), read_ptr_(0), curr_bytes_(0) {}
+
+  /*!
+   * \brief initialize the ecurl request,
+   * \param begin_bytes the beginning bytes of the stream
+   */
+  virtual void InitRequest(size_t begin_bytes, s3url &url) = 0;
+
+ protected:
+  // the total size of the file
+  size_t file_size_ = 0;
+
+  void SetBegin(size_t begin_bytes) {
+    ASSERT_TRUE(begin_bytes < file_size_);
+    curr_bytes_ = begin_bytes;
+  }
+
+ protected:
+  s3url url_;
+
+ private:
+  /*!
+   * \brief called by child class to initialize read
+   * \param begin_bytes the beginning bytes of the stream
+   */
+  void Reset(size_t begin_bytes);
+
+  /*!
+   * \brief try to fill the buffer with at least wanted bytes
+   * \param want_bytes number of bytes we want to fill
+   * \return number of remainning running curl handles
+   */
+
+  int FillBuffer(size_t want_bytes);
+  // data buffer
+  std::string buffer_;
```

Comment:
The buffer here makes it such that these classes cannot be used by more than one thread at a time -- i.e. each thread will have to create its own instance of this class.  Is that the intent?

Reply:
Yeah, that's the intent. But our file stream is used through shared_ptr, so threads will share the same instance.

Reply:
So there's a problem then?  Are you saying that different threads access the same instance?  In that case we need to put in locking or something to ensure that different threads don't compete here.

Reply:
Actually, why do we need a buffer here?

Reply:
because for the normal stream, you should be able to seek. So this one is going to save download from seeking within the same chunk.